vault/ui_custom_message.go

```diff
@@ -0,0 +1,288 @@
+
+const (
+	UICustomMessageKey string = "/custom-messages"
+
+	MaximumCustomMessageCount int = 100
+)
+
+// customMessageBarrierView determines the appropriate logical.Storage to return
+// depending on whether the logical.Storage is being used to access entries for
+// the root namespace or any other namespace based on the provided
+// context.Context.
+func (c *UIConfig) customMessageBarrierView(ctx context.Context) logical.Storage {
+	// If nsBarrierView is nil, which occurs in the non-enterprise edition, then
+	// simply use the barrierStorage.
+	if c.nsBarrierView == nil {
+		return c.barrierStorage
+	}
+
+	// Retrieve the namespace from the context.Context
+	// namespace.FromContext returns an error when:
+	// 1. ctx is nil
+	// 2. there's no Namespace value in ctx
+	// 3. the Namespace value in ctx is nil
+	// In each of those cases, returning the barrierStorage is an appropriate
+	// course of action.
+	ns, err := namespace.FromContext(ctx)
+	if err != nil {
+		return c.barrierStorage
+	}
+
+	return NewBarrierView(c.nsBarrierView, ns.ID)
+}
+
+// ListUICustomMessagesFilters is a struct that captures the different filtering
+// criteria that can be provided to the (*UIConfig).ListCustomMessages method.
+type ListUICustomMessagesFilters struct {
+	authenticated *bool
+	active        *bool
+	messageType   *string
+}
+
+// Authenticated adds the authenticated filter criterion to the receiver
+// ListUICustomMessagesFilters.
+func (f *ListUICustomMessagesFilters) Authenticated(value bool) {
+	f.authenticated = &value
+}
+
+// Active adds the active filter criterion to the receiver
+// ListUICustomMessagesFilters.
+func (f *ListUICustomMessagesFilters) Active(value bool) {
+	f.active = &value
+}
+
+// MessageType adds the messageType filter criterion to the receiver
+// ListUICustomMessagesFilters.
+func (f *ListUICustomMessagesFilters) MessageType(value string) {
+	f.messageType = &value
+}
+
+// UICustomMessageEntry is a struct that contains all of the details of a
+// custom message. This type is used to encode the information into a
+// logical.StorageEntry as well as to transmit custom messages between
+// the logical layer and the request handling layer.
+type UICustomMessageEntry struct {
+	Id            string         `json:"id"`
+	Title         string         `json:"title"`
+	Message       string         `json:"message"`
+	StartTime     time.Time      `json:"start_time"`
+	EndTime       time.Time      `json:"end_time"`
+	Options       map[string]any `json:"options"`
+	Link          map[string]any `json:"link"`
+	Authenticated bool           `json:"authenticated"`
+	MessageType   string         `json:"type"`
+	active        bool
+}
+
+// isTimeNowBetween is a function that determines if the current time, returned
+// by time.Now() is after the provided startTime and before the provided
+// endTime.
+func isTimeNowBetween(startTime, endTime time.Time) bool {
+	now := time.Now()
+
+	return !(startTime.After(now) || endTime.Before(now))
```

Comment:
Consideration: I wonder if we should subtract a little from the start time, i.e. thirty seconds or so, to account for small differences in the sending system versus Vault. The extra wiggle room will make it more unlikely that we create a banner with a start time of 'now' and we need to wait for clocks to sync for it to show up.

Reply:
Interesting point, but this is already so very flaky that I fear doing so would only further complicate things. For instance, the UI will use the LIST endpoint to obtain the set of custom messages. On the backend we will calculate the **active** property, but that calculation is instantly stale. I think that with the current implementation we can at least document it as "**active** is determined at the time that the request was processed by the Vault server"; without having to bring in additional caveats. But this is my point of view and it may very well merit further debate.